realhf/impl/model/backend/inference.py

```diff
@@ -79,63 +79,28 @@ def eval(self):
         return self
 
     @torch.no_grad()
-    def eval_batch(
-        self,
-        input_: SequenceSample,
-        loss_fn: Callable,
-        num_micro_batches: Optional[int] = None,
-    ):
-        if constants.pipe_parallel_world_size() > 1:
-            if num_micro_batches is not None:
-                num_micro_batches = max(
-                    num_micro_batches, self.pipe_runner.default_inf_mbs
-                )
-            return self.pipe_runner.eval_batch(
-                input_=input_,
-                loss_fn=loss_fn,
-                num_micro_batches=num_micro_batches,
-            )
-        else:
-            if num_micro_batches is None:
-                num_micro_batches = 1
-            stat = collections.defaultdict(int)
-            for mb_input in input_.split(num_micro_batches):
-                input_lens = torch.tensor(
-                    flat2d(mb_input.seqlens["packed_input_ids"]),
-                    dtype=torch.int32,
-                    device="cuda",
-                )
-                max_seqlen = int(max(input_lens))
-                cu_seqlens = torch.nn.functional.pad(input_lens.cumsum(0), (1, 0)).int()
-                model_output = self.module(
-                    packed_input_ids=mb_input.data["packed_input_ids"],
-                    cu_seqlens=cu_seqlens,
-                    max_seqlen=max_seqlen,
-                ).logits
-                _, _stat = loss_fn(model_output, mb_input)
-                for k, v in _stat.items():
-                    stat[k] += v
-            return stat
-
     def forward(
         self,
         input_: SequenceSample,
         num_micro_batches: Optional[int] = None,
+        post_hook: Callable[[torch.Tensor, SequenceSample], Any] | None = None,
+        aggregate_fn: Callable[[List[Any]], Any] = torch.cat,
     ):
-        if constants.pipe_parallel_world_size() > 1:
-            if num_micro_batches is not None:
-                num_micro_batches = max(
-                    num_micro_batches, self.pipe_runner.default_inf_mbs
+        # NOTE: Whether to interleave mini-batches in the pipeline results in
+        # similar inference time, but it may introduce a larger memory footprint,
+        # so we split mini-batches in the outer loop and call forward multiple times.
+
+        if num_micro_batches is None:
+            num_micro_batches = 1
+        outputs = []
+        for mb_input in input_.split(num_micro_batches):
+            if constants.pipe_parallel_world_size() > 1:
+                model_output = self.pipe_runner.forward(
+                    input_=input_,
+                    post_hook=post_hook,
+                    aggregate_fn=aggregate_fn,
```

Comment:
The semantic of `num_micro_batches` in `forward` and `generate` is different from the one in `train_batch` now. Maybe we need to unify them, and distinguish `num_micro_batch` here and in pipeline runner.